File: design-pattern-jvm/app/src/test/java/design/pattern/java/patterns/behavioral/observer/ObserverTest.java
package design.pattern.java.patterns.behavioral.observer;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class ObserverTest {
  class BooleanWrapper {
    Boolean value;

    BooleanWrapper(Boolean value) {
      this.value = value;
    }
  }

  @Test
  void shouldNotifyEveryReaderAboutAnyNewsWhenNotifyIsCalled() {
    BooleanWrapper wasReader1Called = new BooleanWrapper(false);
    BooleanWrapper wasReader2Called = new BooleanWrapper(false);
    NewsPaper newsPaper = new NewsPaper();

    newsPaper.subscribe((news) -> {
      wasReader1Called.value = true;
    });
    newsPaper.subscribe((news) -> {
      wasReader2Called.value = true;
    });

    newsPaper.notify(new Notice("test"));
    Assertions.assertTrue(wasReader1Called.value);
    Assertions.assertTrue(wasReader2Called.value);
  }

  @Test
  void shouldUnsubscribeReaderDoNotReceiveNewsWhenReadeWasUnsubscribe() {
    BooleanWrapper wasReader1Called = new BooleanWrapper(false);
    BooleanWrapper wasReader2Called = new BooleanWrapper(false);
    NewsPaper newsPaper = new NewsPaper();

    newsPaper.subscribe((news) -> {
      wasReader1Called.value = true;
    });
    newsPaper.subscribe((news) -> {
      wasReader2Called.value = true;
    }).unsubscribe();

    newsPaper.notify(new Notice("test"));
    Assertions.assertTrue(wasReader1Called.value);
    Assertions.assertFalse(wasReader2Called.value);
  }
}

File: design-pattern-jvm/app/src/test/kotlin/design/pattern/kotlin/patterns/behavioral/observer/ObserverTest.kt
package design.pattern.kotlin.patterns.behavioral.observer

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class ObserverTestKt {
  @Test
  fun shouldNotifyEveryReaderAboutAnyNewsWhenNotifyIsCalled() {
    var wasReader1Called = false
    var wasReader2Called = false
    val newsPaper = NewsPaper()

    newsPaper.subscribe { wasReader1Called = true }
    newsPaper.subscribe { wasReader2Called = true }

    newsPaper.notify(Notice("test"))
    Assertions.assertTrue(wasReader1Called)
    Assertions.assertTrue(wasReader2Called)
  }

  @Test
  fun shouldUnsubscribeReaderDoNotReceiveNewsWhenReadeWasUnsubscribe() {
    var wasReader1Called = false
    var wasReader2Called = false
    val newsPaper = NewsPaper()

    newsPaper.subscribe { wasReader1Called = true }
    newsPaper.subscribe { wasReader2Called = true }.unsubscribe()

    newsPaper.notify(Notice("test"))
    Assertions.assertTrue(wasReader1Called)
    Assertions.assertFalse(wasReader2Called)
  }
}
